python/cudf/cudf/_lib/io/datasource.pxd

```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.
+
+from libcpp.memory cimport unique_ptr
+from cudf._lib.cpp.io.types cimport datasource
+
+cdef class Datasource:
+    cdef unique_ptr[datasource] c_datasource
+
+    cdef inline datasource* get_datasource(self) except *:
```

Comment:
We would need both. Any C++ functions being called within this function should be marked `except +` so that C++ exceptions become Python RuntimeErrors. In addition, since this function returns a non-Python object, it must be marked `except *` for that RuntimeError to be propagated.

Reply:
Ahh, there's `except +*` apparently for functions that can raise both

Reply:
perfect! I was just looking at the docs for something like that. Thanks for sharing.

Reply:
spoke too soon. That seems to cause the cython compiler to crash?

`cudf/_lib/io/datasource.pyx:7:9: Compiler crash in AnalyseDeclarationsTransform`

Reply:
Possible that it's a Cython 3.0 feature then. The alternative is to mark this function `except *`, and all extern C++ functions `except +`.

Reply:
I thought `except *` is just to check if a Python exception occurred, given this is a cdef we have control over, we can safely say we won't be raising a Python exception and just mark it as `except +`, no?

Reply:
> I thought except * is just to check if a Python exception occurred

Yes, and the C++ error gets translated to a Python `RuntimeError`, which we need `except *` to catch properly. If this function returned a Python object we wouldn't have to worry about it

Reply:
Further, even `cdef` functions that don't raise a Python exception explicitly can still do so implicitly (https://github.com/pandas-dev/pandas/pull/33723). So it's always a good idea to mark a Cython function that doesn't return a Python object as `except *`.

Reply:
> > I thought except * is just to check if a Python exception occurred
> 
> Yes, and the C++ error gets translated to a Python `RuntimeError`, which we need `except *` to catch properly. If this function returned a Python object we wouldn't have to worry about it

Does that happen within the function or after the function? I figured the `except +` would catch the C++ exception and the translation to Python exception would then happen outside of the function scope.

Reply:
Here's a small experiment:

```cython
# foo.pyx

cdef extern from * nogil:
    """
    #include <exception>
    void foo() {
        throw std::exception();
    }
    """
    void foo() except +

cdef double c_foo():
    foo()
    return 1.0;

def py_foo():
    c_foo()
```

Compile:

```
$ cythonize -i foo.pyx
```

Note how the `cdef` function `c_foo()` isn't marked `except *`. The exception gets swallowed in this case:


```python
In [1]: from foo import py_foo
In [2]: py_foo()
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
RuntimeError: std::exception
Exception ignored in: 'foo.c_foo'
RuntimeError: std::exception
```

When we mark it `except *` and recompile:

```cython
cdef double c_foo() except *:
    foo()
    return 1.0;
```

the exception is caught correctly:

```python
In [1]: from foo import py_foo
In [2]: py_foo()
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-2-6d1db65baa86> in <module>
----> 1 py_foo()
~/tmp/foo.pyx in foo.py_foo()
     16
     17 def py_foo():
---> 18     c_foo()
     19
~/tmp/foo.pyx in foo.c_foo()
     12
     13 cdef double c_foo() except *:
---> 14     foo()
     15     return 1.0;
     16
RuntimeError: std::exception
```